### PR TITLE
Remove browser-based metadata download

### PR DESCRIPTION
**Pull request by daniel-beck**
Remove browser-based metadata download. [The feature has been deprecated long ago — February 2015 — for security reasons.](https://github.com/jenkinsci/jenkins/blob/6adccd7dda1b894d1dc8927ad3361120c2927548/core/src/main/resources/jenkins/model/DownloadSettings/Warning/message.properties#L24..L25) An administrative monitor (also removed here) recommended against its use.

Per telemetry, less than 2% of instances use this feature.

**Note for Jenkins security team**: See comments on SECURITY-163 that in January/February 2015 said that doing this was the long-term plan. 😅

### Proposed changelog entries

* Major RFE: Remove ability to download update center metadata using the user's browser. Jenkins will no longer inform about available updates without an internet connection. We recommend the use of a local mirror of our update sites, or a self-hosted update center like [Juseppe](https://github.com/jenkinsci/juseppe) in such situations.

### Submitter checklist

- [n/a] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [n/a] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by Wadeck — COMMENTED** [review]
The code removal seems good but I would like to have a plugin alternative for those 2%. Perhaps they have no choice due to their network configuration.

**Comment by daniel-beck** [comment]
> Perhaps they have no choice due to their network configuration.

The alternative is to self-host an update center. This feature never worked great AFAICT, as it informed about metadata changes on the Jenkins UI, but didn't allow for installations.

**Comment by jsoref** [comment]
It seems like that should be what the release note says:

> If your server doesn't have direct access to an update center, you'll need to create a private update center, whether it's a cache or a proxy, in order to support getting updates.

**Comment by daniel-beck** [comment]
@jsoref 

> in order to support getting updates.

This was already the case without a connection: You couldn't actually download updates. Jenkins just informed you that they existed. So it was really a pretty bad setup. Add the security issues it had on top, and it was just garbage.

**Comment by jsoref** [comment]
Hmm. Ok, so, that isn't really the right explanation/description.

> If your server doesn't have the ability to directly contact normal update servers, you have two choices:
> (a) Set up update server(s) (proxy or caching) that it can reach and point it to them.
> (b) Create a script which runs on a computer that has access to update server(s) and jenkins and have it compare and inform you about necessary updates.

**Comment by daniel-beck** [comment]
@jsoref I'd like to keep the changelog brief, I've updated the original comment. The LTS upgrade guide (I expect those users to be much more likely to be on that line anyway) can go into more detail.

File: core/src/main/java/hudson/model/DownloadService.java [hunk 1]
```diff
@@ -72,67 +72,20 @@
...
     /**
      * the prefix for the signature validator name
      */
     private static final String signatureValidatorPrefix = "downloadable";
     /**
      * Builds up an HTML fragment that starts all the download jobs.
+     *
+     * @deprecated browser-based download has been disabled
      */
+    @Deprecated
```
Comment (jglick) [inline core/src/main/java/hudson/model/DownloadService.java]: OK, though I would suggest just deleting the method as I can imagine no reason a plugin might have been calling it.

File: core/src/main/java/hudson/model/DownloadService.java [hunk 2]
```diff
@@ -368,24 +321,6 @@ public JSONObject getData() throws IOException {
             return null;
         }
 
-        /**
-         * This is where the browser sends us the data. 
-         */
-        @RequirePOST
-        public void doPostBack(StaplerRequest req, StaplerResponse rsp) throws IOException {
```
Comment (jglick) [inline core/src/main/java/hudson/model/DownloadService.java]: fix imports

File: core/src/main/java/jenkins/model/DownloadSettings.java [hunk 3]
```diff
@@ -134,18 +109,4 @@ public DailyCheck() {
         }
 
     }
-
-    @Extension public static final class Warning extends AdministrativeMonitor {
-
-        @Override
-        public String getDisplayName() {
-            return Messages.DownloadSettings_Warning_DisplayName();
```
Comment (jglick) [inline core/src/main/java/jenkins/model/DownloadSettings.java]: delete `Messages.properties` entry

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 4]
```diff
@@ -2736,6 +2736,7 @@ function loadScript(href,callback) {
 /**
  * Loads a dynamically created invisible IFRAME.
  */
+// TODO remove this?
```
Comment (jglick) [inline war/src/main/webapp/scripts/hudson-behavior.js]: Yes, I think you can.

File: core/src/main/java/jenkins/model/DownloadSettings.java [hunk 5]
```diff
@@ -134,18 +109,4 @@ public DailyCheck() {
         }
 
     }
-
-    @Extension public static final class Warning extends AdministrativeMonitor {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/DownloadSettings.java]: Forgot to delete `core/src/main/resources/jenkins/model/DownloadSettings/Warning/`.

File: core/src/main/java/jenkins/model/DownloadSettings.java [hunk 6]
```diff
@@ -58,32 +58,10 @@
...
-    public boolean isUseBrowser() {
-        return useBrowser;
-    }
-
-    public void setUseBrowser(boolean useBrowser) {
-        this.useBrowser = useBrowser;
-        save();
-    }
-
     @Override public @Nonnull GlobalConfigurationCategory getCategory() {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/DownloadSettings.java]: There are no longer any actual fields here, so I think `DownloadSettings` could be deleted altogether, and `DailyCheck` moved into its own top-level class or whatever.

Failing that, at least empty out `config.groovy` and delete `help-useBrowser.html`.

**Review by jglick — COMMENTED** [review]
There is more you can delete. Cannot wait for this stuff to be gone. :tada:

**Comment by daniel-beck** [comment]
Finished based on @jglick's review (third commit) and rebased.

File: core/src/main/java/hudson/model/DownloadService.java (hunk 1 above)
Comment (timja) [reply core/src/main/java/hudson/model/DownloadService.java]: +1 on deleting I also can't imagine a plugin calling this =/

Comment (daniel-beck) [reply core/src/main/java/hudson/model/DownloadService.java]: Easy enough to do afterwards, would rather not hold this back longer for that.

File: core/src/main/java/jenkins/DailyCheck.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import org.kohsuke.stapler.HttpResponse;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension
+@Restricted(NoExternalUse.class)
+@Symbol("updateCenterCheck")
+public final class DailyCheck extends AsyncPeriodicWork {
```
Comment (timja) [inline core/src/main/java/jenkins/DailyCheck.java]: For anyone else wondering why this has changed, it used to be in `DownloadSettings`
https://github.com/jenkinsci/jenkins/pull/3970/files#diff-7148b4537324284098f8e7da7de98851L55

**Review by timja — APPROVED** [review]

File: core/src/main/java/jenkins/DailyCheck.java (hunk 7 above)
Comment (daniel-beck) [reply core/src/main/java/jenkins/DailyCheck.java]: Addressed in https://github.com/jenkinsci/jenkins/pull/3970/commits/2bbc99c6a395c102d4387d8ea16c74ad7b0faa4a#diff-bf2e22703bf0c4135b90223a84a3ccaeR46

**Review by Wadeck — APPROVED** [review]
Due to the previous comments, going forward seems reasonable :)
One legacy part less, that's always good.

File: core/src/main/java/jenkins/DailyCheck.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2014 Jesse Glick.
```
Comment (jvz) [inline core/src/main/java/jenkins/DailyCheck.java]: Is this copyright correct?

**Review by jvz — APPROVED** [review]
Everything still seems to work fine. Thanks for taking the initiative to remove this!

**Review by oleg-nenashev — APPROVED** [review]
I agree wit removal of such feature. It is not impossible to reimplement it in a plugin from what I see, but it does not have to be done immediately.

P.S: I am relly surprised 2% of users use it. Unexpectedly many

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback and if somebody fixes CI

**Comment by daniel-beck** [comment]
I kicked CI and it passes now.

> P.S: I am relly surprised 2% of users use it. Unexpectedly many

Legacy default until 2015, dismiss the admin monitor and move on.

**Review by timja — APPROVED** [review]

File: core/src/main/resources/hudson/PluginManager/check.jelly [hunk 9]
```diff
@@ -24,9 +24,8 @@ THE SOFTWARE.
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
 
-  <j:invokeStatic var="ds" className="jenkins.model.DownloadSettings" method="get"/>
   ${%lastUpdated(app.updateCenter.lastUpdatedString)}
-  <f:link href="${ds.useBrowser ? 'checkUpdates' : 'checkUpdatesServer'}" post="true" clazz="yui-button yui-submit-button submit-button primary">
```
Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginManager/check.jelly]: Amended by https://github.com/jenkinsci/jenkins/pull/4613

**Pull request merged**